### PR TITLE
Parameterized CUDA Graph Launch 2

### PR DESCRIPTION
**Pull request by galv**
This is an alternative to cuda graph trees where, instead of copying input values into the static addresses used at stream capture time, we instead simply replace the pointers used by the cuda graph just before calling cudaGraphLaunch().

This is a follow up to #152622

That PR got off to a false start, where I used DWARF debug information to identify the pointers in a cuda graph. This is not going to work anytime soon in production code because we cannot reliably emit dwarf debug symbols for optimized CUDA code.

The current PR uses a much simpler approach to identify the 8-byte aligned 8-byte values that are in fact cuda pointers. First, we know all of the possible input pointers to a cuda kernel because we run on fx graphs, which progamatically have access to all of their inputs and outputs. We can find the start and end of each storage backing these inputs outputs via creating a separate memory pool and calling torch.cuda.memory.memory_snapshot on this memory pool. An 8 byte-aligned 8 byte value is *possibly* a pointer if it, when interpreted as a uint64_t, sits between the start and end of one of these storages. We do stream capture twice. Importantly, we keep the memory pools from the first stream capture alive during the second stream capture. This ensures that no allocation used by the second stream capture will overlap with an allocation used by the stream capture. If the same offset into a parameter is detected as possibly a pointer in both captured graphs, then we declare it to truly be a pointer. Note that there is one flaw in this reasoning: IF the 8 byte value happens to contain padding bytes (which have an undefined value and unfortunately not settable to a specific deterministic value), and the padding bytes are set just right, it is possible that we could incorrectly detect a given 8 byte value as a pointer. This is a very rare situation that I don't think has ever happened, and may basically never ever happen in practice. Maybe we can avoid it by carefully the memory addresses we allocate. But I would prefer to create an optional elf attribute that provides a run-length-encoding of the padding in each argument. That would eliminate the concern.

## Benefits

Parameterized cuda graph launch removes the need to capture each cuda graph to its own memory pool if you want to prevent device memory usage from going out of control. This means that each cuda graph can be rerun on its own without concern for maintaining the invariant that the launch order of a sequence of cuda graphs is equal to the replay order.

We can trivially support input mutations this way.

You no longer have to worry about output tensors being overwritten by other cudagraph replays, since each cuda graph replay will write to a new output tensor, rather than the static output tensor captured at runtime.

## Correctness

Currently, I am running the following tests in torchbench:

```
#!/bin/bash

set -euo pipefail

workdir=trees_vs_params48_release_with_debug_info_dropout_is_all_0_fixed_foreach_copy

# export TORCH_LOGS_FORMAT="%(message)s" 
# export TORCH_LOGS=graph_code,graph_breaks,aot_joint_graph,output_code

if [ -z "${HUGGING_FACE_HUB_TOKEN}" ]; then
  echo "Error: HUGGING_FACE_HUB_TOKEN is not set." >&2
  exit 1
fi

for metric in "accuracy" "performance"; do
    for method in "training" "inference"; do
        for dataset in "huggingface" "torchbench"; do
            mkdir -p $workdir/${dataset}/${method}/${metric}
            time python run_benchmark.py dynamo --speedup-parameterized-cudagraphs-basic --output-directory $workdir/${dataset}/${method}/${metric} --${metric} --${method} --${dataset} 2>&1 | tee $workdir/${dataset}_params_${method}_${metric}.log
        done
    done
done

for metric in "performance" "accuracy"; do
    for method in "training" "inference"; do
        for dataset in "huggingface" "torchbench"; do
            mkdir -p $workdir/${dataset}/${method}/${metric}
            time python run_benchmark.py dynamo --inductor --output-directory $workdir/${dataset}/${method}/${metric} --${method} --${metric} --${dataset} 2>&1 | tee $workdir/${dataset}_trees_${method}_${metric}.log
        done
    done
done
```

The "--accuracy" tests demonstrate that parameterized graph launch works for all tests where cudagraph trees work, so there are no functional/correctness regressions from using parameterized cudagraph launch instead of cudagraph trees.

```
inductor (cudagraph trees):
huggingface, training: 93.48%
torchbench, training: 67.90%
huggingface, inference: 95.65%
torchbench, inference: 73.27%
```

```
parameterized graph launch (me):
huggingface, training: 93.48%
torchbench, training: 67.90%
huggingface, inference: 95.65%
torchbench, inference: 73.27%
```

## Performance

Unfortunately, performance is worse than cudagraph trees right now. The geomean speedup of parameterized cuda graph launch over cudagraph trees (both using the torch.compile inductor backend) on the above benchmark is around 0.95 (I don't remember exactly). This is caused by one reason: I no longer mark inputs as "static". Instead, all inputs are considered to be dynamic, even parameters. Iterating over all of these dynamic inputs to add them to the arguments of the cudagraph before calling cudaGraphLaunch() takes a long amount of time and cannot be hidden. I suppose I don't need to do that. I originally made that change because the "tangents" (i.e., the saved activations of the forward pass) were considered "static" inputs to the backward pass. This is incorrect for parameterized graph launch, where the actual output addresses at replay() time are always new, fresh allocations, rather than whatever the output addresses were at capture time. It is likely that marking parameters as static inputs will be good enough. But I haven't done this yet, partially because I'm not super familiar with how dynamo and inductor decide when particular tensors are "static".

The performance problem is right here: https://github.com/pytorch/pytorch/pull/160351/files#diff-cbda127114b86d36ed63c8b1fabf31722834aba8a857653d88780a1291c5f6b3R481-R506

CC @leijurv, whom I cannot mark as a reviewer for some reason.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: torch/_inductor/utils.py [hunk 1]
```diff
@@ -966,18 +966,20 @@ def get_first_incompatible_cudagraph_node(
         ]
     )
     if torch.are_deterministic_algorithms_enabled():
+        # TODO(galv): Remove these forbidden ops in a separate PR that
```
Comment (galv) [inline torch/_inductor/utils.py:969]: FYI @ngimel your change in https://github.com/pytorch/pytorch/pull/156744 removed an accidental host synchronization: https://github.com/pytorch/pytorch/pull/156744/files#diff-c4e35e76279419d9edda979a995412ee494fad3a97d3fb59486a1f0e326d48beL2212

Therefore, this function which filters out ops that do host synchronization is out of date. Notably, several GNN training and inference workloads in torchbench that did not work with cudagraphs before your change now work with your change.

I have been procrastinating on making a PR for this for a while because, while I would like to actually add appropriate unit tests that the ops in these sets are in fact unsafe for stream capture to a graph, it is a bit hard to do that, since we don't necessarily know suitable arguments to these ops given just their name. You can see chatgpt codex try here: https://chatgpt.com/s/cd_689a4eaa1724819194eec637b9dcf378 But to be clear those tests don't all work because some of the inputs (particularly for the indexing ops) are incorrect.

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 2]
```diff
@@ -140,6 +140,8 @@ void CUDAGeneratorState::register_graph(cuda::CUDAGraph* graph) {
     auto options = at::TensorOptions().device(at::kCUDA).dtype(at::kLong);
     seed_extragraph_ = at::empty({1}, options);
     offset_extragraph_ = at::empty({1}, options);
+    // std::cout << "GALVEZ: seed_extragraph_=" << (uint64_t)seed_extragraph_.data_ptr() << std::endl;
+    // std::cout << "GALVEZ: offset_extragraph_=" << (uint64_t)offset_extragraph_.data_ptr() << std::endl;
```
Comment (eellison) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp]: nit: if DEBUG

File: aten/src/ATen/cuda/CUDAGraph.cpp [hunk 3]
```diff
@@ -1,16 +1,97 @@
...
 #include <ATen/cuda/Exceptions.h>
-#include <ATen/Functions.h>
+#include <ATen/cuda/nvrtc_stub/ATenNVRTC.h>
+#include <c10/core/CPUAllocator.h>
 #include <c10/cuda/CUDACachingAllocator.h>
 #include <c10/cuda/CUDAFunctions.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/cuda/driver_api.h>
+// Not good, aten shouldn't depend upon torch...
+#include <torch/csrc/cuda/CUDAPluggableAllocator.h>
```
Comment (eellison) [inline aten/src/ATen/cuda/CUDAGraph.cpp]: Does this need to be resolved pre blocking ?
